disperser/controller/dispatcher.go

```diff
@@ -0,0 +1,432 @@
+	if err != nil {
+		return fmt.Errorf("failed to aggregate signatures: %w", err)
+	}
+	err = d.blobMetadataStore.PutAttestation(ctx, &corev2.Attestation{
+		BatchHeader:      batchData.Batch.BatchHeader,
+		AttestedAt:       uint64(time.Now().UnixNano()),
+		NonSignerPubKeys: aggSig.NonSigners,
+		APKG2:            aggSig.AggPubKey,
+		QuorumAPKs:       aggSig.QuorumAggPubKeys,
+		Sigma:            aggSig.AggSignature,
+		QuorumNumbers:    nonEmptyQuorums,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to put attestation: %w", err)
+	}
+
+	err = d.updateBatchStatus(ctx, batchData.BlobKeys, v2.Certified)
+	if err != nil {
+		return fmt.Errorf("failed to mark blobs as certified: %w", err)
+	}
+
+	return nil
+}
+
+// NewBatch creates a batch of blobs to dispatch
+// Warning: This function is not thread-safe
+func (d *Dispatcher) NewBatch(ctx context.Context, referenceBlockNumber uint64) (*batchData, error) {
+	blobMetadatas, err := d.blobMetadataStore.GetBlobMetadataByStatus(ctx, v2.Encoded, d.lastUpdatedAt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob metadata by status: %w", err)
+	}
+
+	if len(blobMetadatas) == 0 {
+		return nil, errNoBlobsToDispatch
+	}
+
+	state, err := d.GetOperatorState(ctx, blobMetadatas, referenceBlockNumber)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get operator state: %w", err)
+	}
+
+	keys := make([]corev2.BlobKey, len(blobMetadatas))
+	for i, metadata := range blobMetadatas {
+		if metadata == nil || metadata.BlobHeader == nil {
+			return nil, fmt.Errorf("invalid blob metadata")
+		}
+		blobKey, err := metadata.BlobHeader.BlobKey()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get blob key: %w", err)
+		}
+		keys[i] = blobKey
+		if metadata.UpdatedAt > d.lastUpdatedAt {
+			d.lastUpdatedAt = metadata.UpdatedAt
+		}
+	}
+
+	certs, _, err := d.blobMetadataStore.GetBlobCertificates(ctx, keys)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get blob certificates: %w", err)
+	}
+
+	if len(certs) != len(keys) {
+		return nil, fmt.Errorf("blob certificates not found for all blob keys")
+	}
+
+	certsMap := make(map[corev2.BlobKey]*corev2.BlobCertificate, len(certs))
+	for _, cert := range certs {
+		blobKey, err := cert.BlobHeader.BlobKey()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get blob key: %w", err)
+		}
+
+		certsMap[blobKey] = cert
+	}
+
+	// Keep the order of certs the same as the order of keys
+	for i, key := range keys {
+		c, ok := certsMap[key]
+		if !ok {
+			return nil, fmt.Errorf("blob certificate not found for blob key %s", key.Hex())
+		}
+		certs[i] = c
+	}
+
+	batchHeader := &corev2.BatchHeader{
+		BatchRoot:            [32]byte{},
+		ReferenceBlockNumber: referenceBlockNumber,
+	}
+
+	tree, err := BuildMerkleTree(certs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build merkle tree: %w", err)
+	}
+
+	copy(batchHeader.BatchRoot[:], tree.Root())
```

Comment:
Is it simpler to create the tree first, then populate the `batchHeader` with the root when you create it?

Reply:
I think that's just matter of different order of execution vs. necessarily simpler?